packages/api/src/agent.ts

```diff
@@ -159,23 +159,45 @@ export class AtpAgent {
     try {
       this.session = session
       const res = await this.api.com.atproto.server.getSession()
-      if (!res.success || res.data.did !== this.session.did) {
-        throw new Error('Invalid session')
+      if (res.data.did !== this.session.did) {
+        throw new XRPCError(
+          ResponseType.InvalidRequest,
+          'Invalid session',
+          'InvalidDID',
+        )
       }
       this.session.email = res.data.email
       this.session.handle = res.data.handle
       this.session.emailConfirmed = res.data.emailConfirmed
       this._updateApiEndpoint(res.data.didDoc)
+      this._persistSession?.('update', this.session)
       return res
     } catch (e) {
       this.session = undefined
-      throw e
-    } finally {
-      if (this.session) {
-        this._persistSession?.('create', this.session)
+
+      if (e instanceof XRPCError && e.error) {
+        /*
+         * `ExpiredToken` and `InvalidToken` are handled in
```

Comment:
is `this._refreshSession` implicitly called by the retry logic from calling `getSession`?

Reply:
It's a little confusing, but we override the static `AtpAgent.fetch` here so that if we get an `ExpiredToken` we can try to refresh the session. So calling `getSession` actually calls `this._fetch` here, which calls `this._refreshSession`.

But yeah, basically `expired` will be emit from that handling + afaict we don't get a `ExpiredToken` error from `getSession`.